haas_websocket/main.py

```diff
@@ -0,0 +1,89 @@
+import os
+import json
+import websocket
+import logging
+from haas_websocket.haas_rest_handler import TokenAuth
+from dotenv import load_dotenv
+from google.cloud import pubsub_v1
+
+def restSignIn():
+    rest_agent = TokenAuth()
+    rest_agent.signIn()
+    return rest_agent
+
+def publishMessage(publisher, path, message):
+    future = publisher.publish(path, message)
+
+def filterMessage(jsonData,publisher):
+    project_id = os.getenv('PROJECT_ID')
+    if not project_id:
+        return False
+    
+    encoded = jsonData.encode("utf-8")
+    data = json.loads(jsonData)
+    returnMessage = None
+    path = None
+    published = False
+    
+    data_type = data.get('type')
+    # Filters the messages and publishes messages to specified topic
+    if data_type == 'point':
+        path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('POINT_TOPIC'))
+        returnMessage = 'point'
+    elif data_type == 'heartbeat':
+        path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('HEARTBEAT_TOPIC'))
+        returnMessage = 'heartbeat'
+    elif data_type == 'thing':
+        path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('THING_TOPIC'))
+        returnMessage = 'thing'
+    elif data_type == 'location':
+        path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('LOCATION_TOPIC'))
+        returnMessage = 'location'
+    elif data_type is None:
+        returnMessage = 'keepAlive'
+    if path and returnMessage:
+        publishMessage(publisher,path,encoded)
+        published = True
+
+    
+    return returnMessage, published
+
+def running():
+    return True
+
+def startWebsocket(publisher):
+    rest = restSignIn()
+    b_token = rest.getToken()
+    ws_endpoint = os.getenv('HAAS_WSS_ENDPOINT')
+    ws = websocket.create_connection(ws_endpoint+b_token)
+    while running() == True:
+        check_pass = rest.checkPassword()
+        check_token = rest.checkToken()
+        if check_pass == False:
+            ws.close()
+            rest.passwordUpdate() # updates local token from the secret manager
+            b_token = rest.signIn()
+            ws = websocket.create_connection(ws_endpoint+b_token) # restarts the websocket
```

Comment:
Yeah, the ws_endpont (HAAS_API_ENDPOINT env variable) is expected to end in a / for the b_token to proceed it. I can mention that in the readme.

Reply:
Yeah, let's add that to the README so we don't run into troubleshooting issues down the road.

Reply:
Did this get added into the README? I'm not seeing it in there.

Reply:
yeah I added this to the Environmental Setup section

"In the .env file, all endpoints must end with a backslash (ex: www.google.com/) following the url otherwise the program will not parse the api endpoint or websocket variables properly."